src/parser/uniterm.rs

```diff
@@ -43,35 +181,55 @@ impl UniRecord {
                 self.tail
                     .map(|(tail, _)| tail)
                     .unwrap_or(Types(AbsType::RowEmpty())),
-                |acc, (path_elem, rt)| {
-                    match path_elem {
-                        FieldPathElem::Ident(id) => {
-                            // At parsing stage, all `Rc`s must be 1-counted. We can thus call
-                            // `into_owned()` without risking to actually clone anything.
-                            match rt.term.into_owned() {
-                                Term::MetaValue(MetaValue {
-                                    doc: None,
-                                    types: Some(ctrt),
-                                    contracts,
-                                    priority: MergePriority::Normal,
-                                    value: None,
-                                }) if contracts.is_empty() => Ok(Types(AbsType::RowExtend(
-                                    id,
-                                    Some(Box::new(ctrt.types)),
-                                    Box::new(acc),
-                                ))),
-                                _ => {
-                                    // Position of identifiers must always be set at this stage
-                                    // (parsing)
-                                    let span_id = id.pos.unwrap();
-                                    let term_pos = rt.pos.into_opt().unwrap_or(span_id);
-                                    Err(InvalidRecordTypeError(TermPos::Original(
-                                        RawSpan::fuse(span_id, term_pos).unwrap(),
-                                    )))
+                |acc, (mut path, rt)| {
+                    // We don't support compound paths for types, yet.
+                    if path.len() > 1 {
+                        let span = path
+                            .into_iter()
+                            .map(|path_elem| match path_elem {
+                                FieldPathElem::Ident(id) => id.pos.into_opt(),
+                                FieldPathElem::Expr(rt) => rt.pos.into_opt(),
+                            })
+                            .reduce(|acc, pos| {
+                                acc.zip(pos)
+                                    .and_then(|(acc, span)| RawSpan::fuse(acc, span))
+                            })
+                            .flatten();
+
+                        Err(InvalidRecordTypeError(
+                            span.map_or(TermPos::None, TermPos::Original),
+                        ))
+                    } else {
+                        let elem = path.pop().unwrap();
```

Comment:
It's true that it's currently not statically enforced, but the path is the LHS of a field declaration `foo.bar.[...] = value`, and it can't be empty in the grammar, so the parser should never outputs empty paths. On one side we could make sure this code never panics by changing the condition from `path.len() > 1` to `path.len() != 1`, and everything would work fine. On the other hand, it's an illegal state, so I'm tempted to keep it like this and let it fail hard if this invariant were to be violated.